### PR TITLE
Support calculating normals for indexed meshes

### PR DESCRIPTION
**Pull request by tygyh**
# Objective

- Finish #3987

## Solution

- Rebase and fix typo.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 1]
```diff
@@ -536,15 +536,16 @@ impl Mesh {
...
-    pub fn compute_flat_normals(&mut self) {
-        assert!(self.indices().is_none(), "`compute_flat_normals` can't work on indexed geometry. Consider calling `Mesh::duplicate_vertices`.");
-
+    /// Panics if [`Mesh::ATTRIBUTE_POSITION`] is not of type `float3`.
+    /// Panics if the mesh has any other topology than [`PrimitiveTopology::TriangleList`].
+    ///
+    /// FIXME: The should handle more cases since this is called as a part of gltf
+    /// mesh loading where we can't really blame users for loading meshes that might
+    /// not conform to the limitations here!
+    pub fn compute_normals(&mut self) {
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/mesh/mesh/mod.rs:545]: Can you undo the rename here to avoid a breaking change?

**Comment by robtfm** [comment]
how does this handle something like a 2-sided quad, where the shared face normals are exact opposites? i think the normal will be NaN, but i'm not sure if that's a big problem. if it is an issue you can replace the `normalize` with `normalize_or_zero`.

nit: i would personally use `chunks_exact` (like in the None branch) rather than iterating individually and branching on each 3rd entry.

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 2]
```diff
@@ -553,18 +554,56 @@ impl Mesh {
...
+                }
+
+                // average (smooth) normals for shared vertices...
+                // TODO: support different methods of weighting the average
+                for i in 0..normals.len() {
+                    let count = adjacency_counts[i];
+                    if count > 0 {
+                        normals[i] = (Vec3::from(normals[i]) / (count as f32)).normalize().into();
+                    }
+                }
```
Comment (pcwalton) [inline crates/bevy_render/src/mesh/mesh/mod.rs:588]:
```suggestion
                let mut normals = vec![[0.0; 3]; positions.len()];
                let mut adjacency_counts = vec![0; positions.len()];

                for face in indices.chunks(3) {
                    let normal = face_normal(positions[face[0]], positions[face[1]], positions[face[2]]);
                    for &index in &face {
                        normals[index as usize] = (Vec3A::from(normal) + Vec3A::from(normals[corner])).into();
                        adjacency_counts[index as usize] += 1;
                    }
                }

                // Average (smooth) normals for shared vertices.
                // TODO: Weight by angle.
                for (normal, &count) in normals.iter_mut().zip(adjacency_counts.iter()) {
                    if count > 0 {
                        *normal = (Vec3A::from(*normal) / count as f32).normalize().into();
                    }
                }
```

**Comment by pcwalton** [comment]
One suggestion to use more idiomatic iterators, otherwise LGTM.

**Comment by superdump** [comment]
Swinging through - does mikktspace have anything for calculating vertex normals? Or else, how do people normally calculate vertex normals in an automated way in industry software? What does Blender / UE / Unity / Godot do?

File: crates/bevy_render/src/mesh/mesh/mod.rs (hunk 2 above)
Comment (tygyh) [reply crates/bevy_render/src/mesh/mesh/mod.rs:588]: > for face in indices.chunks(3) {

There does not seem to be a method called chunks in Indices and I do not know how to take 3 at a time in some other way.

**Comment by IceSentry** [comment]
This PR should probably be reverted. LIke I said in the original PR using `compute_flat_normals()` here is misleading because it's not computing flat normals. We should have a separate `compute_smooth_normals()` function for this.

**Comment by IceSentry** [comment]
@tygyh if you extract this to a separate function that properly indicates that it computes smooth normals and warns that it might break normal maps then I'll approve.

You could also add a warning in compute_flat_normals to use that function when using an indexed mesh.

**Comment by ManevilleF** [comment]
I did a similar implentation in [bevy_silk](https://github.com/ManevilleF/bevy_silk/blob/main/src/components/cloth_rendering.rs#L187) for flat and smooth normals, using `chunk_exact` if that's useful

File: crates/bevy_render/src/mesh/mesh/mod.rs (hunk 2 above)
Comment (rlidwka) [reply crates/bevy_render/src/mesh/mesh/mod.rs:588]: `chunks` is a method on any iterator: https://doc.rust-lang.org/std/primitive.slice.html

should probably use `chunks_exact` here though

**Pull request merged**